File: LeetCodes/open-the-lock/4-23-2024, 3_03_43 PM/open-the-lock.java
// Link: https://leetcode.com/problems/open-the-lock

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: open-the-lock
 *  RUNTIME: 106 ms
 *  MEMORY: 46.6 MB
 *  DATE: 4-23-2024, 3:03:43 PM
 *
 */

// Solution:

class Solution {
    public int openLock(String[] deadends, String target) {
        Queue<String> pq = new LinkedList<>();
        Set<String> vis = new HashSet<>();

        Set<String> dead = new HashSet<>();
        dead.addAll(Arrays.asList(deadends));

        String start = "0000";

        if(dead.contains(target) || dead.contains(start))
            return -1;

        pq.add("0000");
        vis.add("0000");

        int steps = 0;
        while(!pq.isEmpty()) {
            int size = pq.size();
            while(size > 0) {
                String str = pq.remove();

                if(str.equals(target)) {
                    return steps;
                }
                
                List<String> neighbours = generate_neighbours(str);
                for(String nextStr: neighbours) {
                    if(dead.contains(nextStr) || vis.contains(nextStr))
                        continue;
                    pq.add(nextStr);
                    vis.add(nextStr);
                }

                size--;
            }
            steps++;
        }
        return -1;
    }

    private Pair<String,Integer> make_pair(String str, Integer num) {
        return new Pair<>(str, num);
    }

    private List<String> generate_neighbours(String str) {
        List<String> neighbours = new ArrayList<>();
        char[] chars = str.toCharArray();

        for(int i=0; i<4; i++) {
            char orgChar = chars[i];
            chars[i] = (orgChar == '0') ? '9' : (char)(orgChar - 1);
            neighbours.add(new String(chars));

            chars[i] = (orgChar == '9') ? '0' : (char)(orgChar + 1);
            neighbours.add(new String(chars));

            chars[i] = orgChar;
        }

        return neighbours;
    }
}

File: LeetCodes/open-the-lock/4-23-2024, 3_14_11 PM/open-the-lock.cpp
// Link: https://leetcode.com/problems/open-the-lock

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: open-the-lock
 *  RUNTIME: 287 ms
 *  MEMORY: 140.8 MB
 *  DATE: 4-23-2024, 3:14:11 PM
 *
 */

// Solution:

class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        queue<string> q;
        unordered_set<string> vis, dead(deadends.begin(), deadends.end());

        string start = "0000";

        if(dead.count(start) || dead.count(target))
            return -1;

        q.push(start);
        vis.insert(start);

        int steps = 0;
        while(!q.empty()) {
            int size = q.size();

            while(size--) {
                string str = q.front();
                q.pop();

                if(str == target)
                    return steps;

                vector<string> neighbours = generate_neighbours(str); 
                for(auto nextStr: neighbours) {
                    if(vis.count(nextStr) || dead.count(nextStr))
                        continue;
                    
                    q.push(nextStr);
                    vis.insert(nextStr);
                }
            }
            steps++;
        }

        return -1;
    }

    vector<string> generate_neighbours(string str) {
        vector<string> neighs;

        for(int i=0; i<4; i++) {
            char orgChar = str[i];

            str[i] = (orgChar == '0') ? '9' : (char)(orgChar - 1);
            neighs.push_back(str);

            str[i] = (orgChar == '9') ? '0' : (char)(orgChar + 1);
            neighs.push_back(str);

            str[i] = orgChar;
        }

        return neighs;
    }
};
